lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -1005,6 +1007,62 @@ def invocation_step(self, trans, invocation_id, step_id, **kwd):
         )
         return self.__encode_invocation_step(trans, invocation_step)
 
+    def _generate_invocation_metrics(self, trans, invocation_id, **kwd):
+        """
+        build a dictionary of invocation metrics
+        """
+        decoded_workflow_invocation_id = self.decode_id(invocation_id)
+        workflow_invocation = self.workflow_manager.get_invocation(trans, decoded_workflow_invocation_id)
+        history = workflow_invocation.history
+
+        metrics, h_metrics = {}, {}
+        h_contents = self.history_contents_manager.contained(history)
+
+        for i, step in enumerate(workflow_invocation.steps):
+            if step.workflow_step.type == 'tool':
+                tool_id = step.workflow_step.tool_id.split('/')[-2:]
+                for job in step.jobs:
+                    job_metrics = summarize_job_metrics(trans, job)
+                    metrics[i] = {
+                        'tool_name': tool_id[0],
+                        'tool_version': tool_id[1],
+                        'galaxy_slots': '',
+                        'runtime_seconds': '',
+                        'job_inputs': [],
+                        'job_outputs': []
+                    }
+                    for value in [x for x in job_metrics if x['name'] == 'galaxy_slots']:
+                        metrics[i]['galaxy_slots'] = value['raw_value']
+                    for value in [x for x in job_metrics if x['name'] == 'runtime_seconds']:
+                        metrics[i]['runtime_seconds'] = value['raw_value']
+                    for job_input in job.input_datasets:
+                        if hasattr(job_input.dataset, 'dataset_id'):
+                            for h in [x for x in h_contents if x.dataset_id == job_input.dataset.id]:
+                                h_metrics = {
+                                    'total_size': int(h.dataset.total_size),
+                                    'uuid': str(h.dataset.uuid),
+                                    'job_input_name': job_input.name
+                                }
+                                metrics[i]['job_inputs'].append(h_metrics)
```

Comment:
You don't need the history contents at all here:
```suggestion
                        if job_input.dataset.dataset:
                            dataset = job_input.dataset
                            dataset_metrics = {
                                'total_size': int(dataset.total_size),
                                'id': trans.security.encode_id(dataset.id),
                                'model_class': dataset.__class__.__name__,
                                'job_input_name': job_input.name
                            }
                            metrics[i]['job_inputs'].append(dataset_metrics)
```
I'd use the item id instead of the uuid. The id is the primary identifier we use in Galaxy, with this id you can get more info about this dataset for instance ... we don't support querying by uuid for all api endpoints. The input might also be a LIbraryDatasetAssociation, so that's why I'd also dump the model_class into the metrics.

You can use the same technique for job outputs.